### PR TITLE
Try and clean up writeFileAllUnseekable by using readers

### PR DESCRIPTION
**Pull request by daurnimator**
`writeFileAll*` may be beyond help (#7014) but I tried anyway...

### CODE REVIEW

File: lib/std/io/early_eof_reader.zig [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+// Copyright (c) 2015-2020 Zig Contributors
+// This file is part of [zig](https://ziglang.org/), which is MIT licensed.
+// The MIT license requires this copyright notice to be included in all copies
+// and substantial portions of the software.
+const std = @import("../std.zig");
+const io = std.io;
+const assert = std.debug.assert;
+const testing = std.testing;
+
+pub fn EarlyEOFReader(comptime ReaderType: type) type {
```
Comment (LemonBoy) [inline lib/std/io/early_eof_reader.zig:11]: Oh `EarlyEOFReader` may be really useful, I like it.
The name doesn't really convey what it does though, I'd go with `LimitedReader` (name borrowed from [golang](https://golang.org/pkg/io/#LimitedReader) or something along that line.

File: lib/std/io/early_eof_reader.zig [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    return struct {
+        child_stream: ReaderType,
```
Comment (LemonBoy) [inline lib/std/io/early_eof_reader.zig:13]: Nit, this is a reader and not a stream so `inner_reader` may be a better name.

**Review by LemonBoy — APPROVED** [review]
Gotta admit that `pump` was a really nice touch.

File: lib/std/io/early_eof_reader.zig (hunk 2 above)
Comment (daurnimator) [reply lib/std/io/early_eof_reader.zig:13]: FWIW I copied the field name from `CountingWriter`.

Comment (LemonBoy) [reply lib/std/io/early_eof_reader.zig:13]: I didn't write/review that code :^), I'm [sticking](https://github.com/ziglang/zig/blob/2b7781d82ad8d2234b89257676670957e005f214/lib/std/compress/zlib.zig#L27) to this naming convention.
My understanding is that a `Stream` is the backing object out of which you pull a `Reader` or a `Writer`.

**Comment by LemonBoy** [comment]
Oh well, @joachimschmidt557 changed the `num_bytes` parameter to `usize` in  `io.Reader.skipBytes`...
I hope you don't mind changing it back to `u64` (file offsets have nothing to do with the pointer width) to fix the red CI.

**Comment by joachimschmidt557** [comment]
@LemonBoy oops, sorry. I think I used `usize` because I slice the temporary buffer with `amt`, also a `usize`. `amt` is the minimum of the remaining bytes (currently also `usize`) and the buffer size. Changing `num_bytes` to `usize` would require an `@intCast` somewhere as we can't change everything to `u64` (because of the slicing).

**Comment by LemonBoy** [comment]
`num_bytes` is an offset and must be `u64`, the temporary buffer size can stay as `usize` as that's the size of an object in memory.
You've defined `amt` as `std.math.min(remaining, options.buf_size)`, the only change needed is to turn `num_bytes` and `remaining` into `u64`, the slicing amount will surely fit into a `usize` because `buf_size` _is_ a `usize`.

**Comment by joachimschmidt557** [comment]
@LemonBoy oops, yes, I forgot that the return type of `std.math.min` will be `usize` in that case.

File: lib/std/io/early_eof_reader.zig (hunk 1 above)
Comment (daurnimator) [reply lib/std/io/early_eof_reader.zig:11]: hrm. I agree the name isn't great, but I don't think that `LimitedReader` is better...

Open to other suggestions.

**Comment by daurnimator** [comment]
> `num_bytes` is an offset and must be `u64`

added to this PR. Though I noticed something else terrible!:

```
lib/std/os/bits/linux/sparcv9.zig:350:pub const off_t = isize;
lib/std/os/bits/linux/arm64.zig:422:pub const off_t = isize;
lib/std/os/bits/linux/riscv64.zig:366:pub const off_t = isize;
```

**Comment by LemonBoy** [comment]
> Though I noticed something else terrible!:

Awesome! [Posix](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_types.h.html) mandates `off_t` to be signed so we're in for some fun!

File: lib/std/io/early_eof_reader.zig (hunk 1 above)
Comment (frmdstryr) [reply lib/std/io/early_eof_reader.zig:11]: Maybe some variant of `MaxLengthReader`, `FixedSizeReader`, or something idk.. 

EarlyEOFReader was really confusing before reading what it does.

Comment (frmdstryr) [reply lib/std/io/early_eof_reader.zig:11]: Or `CappedReader`

**Comment by andrewrk** [comment]
Thanks, landed in 483c057a771aca3b045ec8148fa5fb1a5e28fa17. Went with @LemonBoy's suggestion to follow the Go convention for the name.

**Pull request closed without merging**